trillian/integration/ct_integration.go

```diff
@@ -589,6 +593,178 @@ func RunCTIntegrationForLog(cfg *configpb.LogConfig, servers, metricsServers, te
+	// at least 2000 so the queue builds up a bit.
+	atLeast := 2000
+	count := atLeast + rand.Intn(3000-atLeast)
+	fmt.Printf("%s: Starting upload of %d certificates ....\n", t.prefix, count)
+	for i := 1; i <= count; i++ {
+		chain, err := makeCertChain(leafChain, leafCert, caCert, signer, time.Now().Add(24*time.Hour))
+		if err != nil {
+			return err
+		}
+		_, err = t.client().AddChain(ctx, chain)
+		t.stats.done(ctfe.AddChainName, 200)
+		if err != nil {
+			return fmt.Errorf("got AddChain(int-ca.cert)=(nil,%v); want (_,nil)", err)
+		}
+	}
+	fmt.Printf("%s: Upload of %d certificates complete\n", t.prefix, count)
+
+	// Stage 3: Set the log to DRAINING using the admin server.
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	if err := setTreeState(ctx, t.adminServer, t.cfg.LogId, trillian.TreeState_DRAINING); err != nil {
+		return err
+	}
+
+	// Stage 4a: Get an updated STH. We'll use this point for a consistency
+	// proof later.
+	sth1, err := t.client().GetSTH(ctx)
+	t.stats.done(ctfe.GetSTHName, 200)
+	if err != nil {
+		return fmt.Errorf("got GetSTH()=(nil,%v); want (_,nil)", err)
+	}
+	if sth1.Version != 0 {
+		return fmt.Errorf("sth.Version=%v; want V1(0)", sth1.Version)
+	}
+	fmt.Printf("%s: Got STH(time=%q, size=%d): roothash=%x\n", t.prefix, timeFromMS(sth1.Timestamp), sth1.TreeSize, sth1.SHA256RootHash)
+
+	// Stage 4b: Wait for the queue to drain and everything to be integrated.
+	sth2, err := t.awaitTreeSize(context.Background(), uint64(count), true, mmd)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("%s: Got STH(time=%q, size=%d): roothash=%x\n", t.prefix, timeFromMS(sth2.Timestamp), sth2.TreeSize, sth2.SHA256RootHash)
+
+	// Stage 5. Get a consistency proof from sth1 to sth2 and verify it.
+	proof, err := t.client().GetSTHConsistency(ctx, sth1.TreeSize, sth2.TreeSize)
+	t.stats.done(ctfe.GetSTHConsistencyName, 200)
+	if err != nil {
+		return err
+	}
+	if err := checkCTConsistencyProof(sth1, sth2, proof); err != nil {
+		return err
+	}
+	fmt.Printf("%s: VerifiedConsistency(time=%q, size1=%d, size2=%d): final roothash=%x\n", t.prefix, timeFromMS(sth2.Timestamp), sth1.TreeSize, sth2.TreeSize, sth2.SHA256RootHash)
+
+	// Stage 6. Try to submit a chain and it should be rejected with 403.
+	var chain [21][]ct.ASN1Cert
+	chain[0], err = GetChain(testdir, "int-ca.cert")
+	if err != nil {
+		return fmt.Errorf("failed to load certificate: %v", err)
+	}
+	_, err = t.client().AddChain(ctx, chain[0])
+	t.stats.done(ctfe.AddChainName, 403)
+	if err == nil || !strings.Contains(err.Error(), "403") {
+		return fmt.Errorf("got AddChain(int-ca.cert)=(nil,%v); want (_,err inc. 403)", err)
+	}
+
+	// Stage 7 - Set the log to FROZEN using the admin server.
+	ctx, cancel = context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	if err := setTreeState(ctx, t.adminServer, t.cfg.LogId, trillian.TreeState_FROZEN); err != nil {
+		return err
```

Comment:
Should these raw `return err` be decorated as elsewhere so we can tell where it went wrong if something fails?

Reply:
Done.